File: 46-permutations/46-permutations.cpp
class Solution {
private:
    vector<int> reset(vector<int> vec, int index){
        vector<int> res(vec.size() - 1 );
        int t = 0;
        for(int i=0; i<vec.size(); i++){
            if(i != index){
                res[t] = vec[i];
                t++;
            }
        }
        return res;
    }

public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        
        // BASE CASE
        if(nums.size() == 0){
            vector<int> sample;
            result.push_back(sample);
            return result;
        }
        
        // SMALL PROBLEM
        for(int i=0; i<nums.size(); i++){
            int getElement{nums[i]};
            vector<int> newVec = reset(nums, i);
            vector<vector<int>> returnedVector = permute(newVec);
            for(vector<int> vec: returnedVector){
                vec.push_back(getElement);
                result.push_back(vec);
            }
        }
        return result;
    }
};

File: 46-permutations/46-permutations.java
class Solution {
    static int [] reset(int[] arr, int index){
        int [] res = new int[arr.length - 1];
        int t = 0;
        for(int i=0; i<arr.length; i++){
            if(i != index){
                res[t] = arr[i];
                t++;
            }
        }
        return res;
    }
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        // BASE CASE
        if(nums.length == 0){
            result.add(new ArrayList<Integer>());
            return result;
        }
        
        // SMALL PROBLEM
        for(int i=0; i<nums.length; i++){
            int getElement = nums[i];
            int []newArr = reset(nums, i);
            List<List<Integer>> returnedList = permute(newArr);
            for(List<Integer> t: returnedList){
                t.add(getElement);
                result.add(t);
            }
        }
        return result;
    } 
}